File: leetcode/insert-interval.py
# https://leetcode.com/problems/insert-interval/

# Tags: Array

import timeit
from typing import List


# First check if we need to insert before all intervals
# Then iterate over the elements of intervals and for each element check if we need to insert before or merge
# Once we are done, check if we need to insert after the whole array
#
# Time complexity: O(n) we check all elements once
# Space complexity: O(n) we use a list to build the result set
#
# Runtime: 159 ms, faster than 17.84% of Python3 online submissions for Insert Interval.
# Memory Usage: 17.2 MB, less than 92.25% of Python3 online submissions for Insert Interval.
class IterativeOneArray:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if not intervals:
            return [newInterval]

        result = []
        insert_start, insert_end = newInterval

        # Check if we need to insert before all elements
        if insert_end < intervals[0][0]:
            result.append(newInterval)

        for start, end in intervals:
            # Check if we need to insert between the previous element and the current element without merging
            if result and result[-1][1] < insert_start and insert_end < start:
                result.append(newInterval)
            # Check if we don't need to do anything
            if end < insert_start or start > insert_end:
                # Just insert this value
                result.append([start, end])
            # We need to merge
            else:
                # Merge the insert element with this element
                merged_start = min(start, insert_start)
                merged_end = max(end, insert_end)

                # Check if we need to merge with the previous element
                if result and result[-1][1] >= merged_start:
                    result[-1] = [result[-1][0], max(merged_end, result[-1][1])]
                # If result is empty, or the previous element's end is strictly before the merged start, just insert this element
                else:
                    result.append([merged_start, merged_end])

        # Check if we need to insert after all elements
        if result[-1][1] < insert_start:
            result.append(newInterval)

        return result


# Iterate over the elements but use newInterval to create the merged interval instead of using multiple conditionals.
#
# Time complexity: O(n) we check all elements once
# Space complexity: O(n) we use a list to build the result set
#
# Runtime: 87 ms, faster than 89.89% of Python3 online submissions for Insert Interval.
# Memory Usage: 17.4 MB, less than 53.14% of Python3 online submissions for Insert Interval.
class IterUpdateInsert:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []

        for i, int in enumerate(intervals):

            # If we can insert newInterval before the current interval, do it and return the result
            if newInterval[1] < int[0]:
                res.append(newInterval)
                return res + intervals[i:]

            # If the end of the current interval does not overlap the insert interval, append the current interval
            elif newInterval[0] > int[1]:
                res.append(int)

            # If there is overlap, start building the insert interval by merging the original one with the overlapping ones
            # but do not insert it yet
            else:
                newInterval = [min(newInterval[0], int[0]), max(newInterval[1], int[1])]

        # If the code gets here, newInterval could not fit before an existing interval and has not been inserted yet
        res.append(newInterval)
        return res


def test():
    executors = [IterativeOneArray, IterUpdateInsert]
    tests = [
        [[[1, 3], [6, 9]], [2, 5], [[1, 5], [6, 9]]],
        [[[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]], [4, 8], [[1, 2], [3, 10], [12, 16]]],
        [[[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]], [2, 3], [[1, 5], [6, 7], [8, 10], [12, 16]]],
        [[], [1, 2], [[1, 2]]],
        [[[1, 5]], [6, 8], [[1, 5], [6, 8]]],
        [[[6, 8]], [1, 5], [[1, 5], [6, 8]]],
        [[[0, 4], [6, 8]], [1, 5], [[0, 5], [6, 8]]],
        [[[3, 5], [12, 15]], [6, 6], [[3, 5], [6, 6], [12, 15]]],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(int(float("1e4"))):
            for i, t in enumerate(tests):
                sol = executor()
                result = sol.insert(t[0], t[1])
                exp = t[2]
                assert (
                    result == exp
                ), f"\033[93m» {result} <> {exp}\033[91m for test {i} using \033[1m{executor.__name__}"
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        res = "{0:20}{1:10}{2:10}".format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/insert-interval.rs
// 57. Insert Interval
// 🟠 Medium
//
// https://leetcode.com/problems/insert-interval/
//
// Tags: Array

#[derive(Debug, Clone)]
struct Interval {
    start: i32,
    end: i32,
}

impl Interval {
    fn merge(&self, other: &Interval) -> Self {
        Interval {
            start: self.start.min(other.start),
            end: self.end.max(other.end),
        }
    }
}

struct Solution;
impl Solution {
    /// Iterate over the intervals in the input. Keep a mutable copy of the new interval, insert
    /// all intervals that come fully before the new interval, then merge the overlapping intervals
    /// with the new interval by mutating it but not inserting it yet. Once we intert in, append
    /// the remaining elements and return the result. If we exit the loop, we have not inserted the
    /// new interval yet, do it and return the result.
    ///
    /// Time complexity: O(n) - We iterate over the n elements in the input and do constant time
    /// work for each.
    /// Space complexity: O(1) - If we don't take into account the result vector.
    ///
    /// Runtime 2 ms Beats 73%
    /// Memory 2.65 MB Beats 51%
    #[allow(dead_code)]
    pub fn insert_1(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut res = vec![];
        let mut new_interval = new_interval;
        for (idx, int) in intervals.iter().enumerate() {
            if new_interval[1] < int[0] {
                res.push(new_interval);
                res.extend(intervals[idx..].into_iter().map(|v| v.clone()));
                return res;
            } else if new_interval[0] > int[1] {
                res.push(int.to_vec());
            } else {
                new_interval = vec![new_interval[0].min(int[0]), new_interval[1].max(int[1])];
            }
        }
        res.push(new_interval);
        res
    }

    /// Same logic as the previous solution but use a struct and impl to make it cleaner. I
    /// expected this solution to be slower because it iterates the input vector extra times to map
    /// the input to a Vec<Interval> and then map the result back to Vec<Vec<_>> but it got the
    /// sames times as the previous result.
    ///
    /// Time complexity: O(n) - We iterate over the n elements in the input and do constant time
    /// work for each.
    /// Space complexity: O(1) - If we don't take into account the result vector.
    ///
    /// Runtime 2 ms Beats 73%
    /// Memory 2.65 MB Beats 51%
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let intervals = intervals
            .iter()
            .map(|i| Interval {
                start: i[0],
                end: i[1],
            })
            .collect::<Vec<_>>();
        let mut res: Vec<Interval> = vec![];
        let mut new_interval = Interval {
            start: new_interval[0],
            end: new_interval[1],
        };
        for (idx, int) in intervals.iter().enumerate() {
            if new_interval.end < int.start {
                res.push(new_interval);
                res.extend(intervals[idx..].into_iter().map(|v| v.clone()));
                return res.iter().map(|i| vec![i.start, i.end]).collect::<Vec<_>>();
            } else if new_interval.start > int.end {
                res.push(int.clone());
            } else {
                new_interval = new_interval.merge(&int);
            }
        }
        res.push(new_interval);
        res.iter().map(|i| vec![i.start, i.end]).collect::<Vec<_>>()
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![], vec![5, 7], vec![[5, 7]]),
        (vec![[1, 3], [6, 9]], vec![2, 5], vec![[1, 5], [6, 9]]),
        (
            vec![[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]],
            vec![4, 8],
            vec![[1, 2], [3, 10], [12, 16]],
        ),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::insert(
            t.0.iter().map(|arr| arr.to_vec()).collect::<Vec<_>>(),
            t.1.clone(),
        );
        let expected = t.2.iter().map(|arr| arr.to_vec()).collect::<Vec<_>>();
        if res == expected {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {:?}!!\x1b[0m",
                i, expected, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
